pandas/core/reshape/merge.py

```diff
@@ -1648,16 +1648,8 @@ def get_result(self) -> DataFrame:
         right_join_indexer: np.ndarray | None
 
         if self.fill_method == "ffill":
-            # error: Argument 1 to "ffill_indexer" has incompatible type
-            # "Optional[ndarray]"; expected "ndarray"
-            left_join_indexer = libjoin.ffill_indexer(
-                left_indexer  # type: ignore[arg-type]
-            )
-            # error: Argument 1 to "ffill_indexer" has incompatible type
-            # "Optional[ndarray]"; expected "ndarray"
-            right_join_indexer = libjoin.ffill_indexer(
-                right_indexer  # type: ignore[arg-type]
-            )
+            left_join_indexer = libjoin.ffill_indexer(np.asarray(left_indexer))
```

Comment:
Please avoid calling asarray. This adds unnecessary overhead and does not guarantee that we get an array back if None is passed into it.

You can use assert if it is guaranteed, that left_indexer is not None here. Adjusting the return type of _get_join_info would be even better

Reply:
@phofl Hi Patrick. I see what you mean about the overhead of ```np.asarray()```. None seems to come from L957 and L962 in ```_get_join_info``` which comes from return value of ```_left_join_on_index``` L2071.  Not sure what a good alternative np.ndarray alternative to None would be?

Reply:
can we reason about the code s.t. we know we can _never_ get here with left_indexer being None?  if so, then can do a `left_indexer = cast(np.ndarray, left_indexer)`.  otherwise

```
if left_indexer is None:
     raise AppropriateException("Useful message")
```

Reply:
@jbrockmendel thank you for the suggestion.  I was unsure if it was okay to use ```typing.cast()```  Do these changes look okay?